src/clr-addons/daterangepicker/providers/daterange-parsing.service.ts

```diff
@@ -0,0 +1,253 @@
+import { FormatWidth, getLocaleDateFormat } from '@angular/common';
+import { Inject, Injectable, LOCALE_ID } from '@angular/core';
+
+import {
+  BIG_ENDIAN,
+  DELIMITER_REGEX,
+  InputDateDisplayFormat,
+  LITTLE_ENDIAN,
+  LITTLE_ENDIAN_REGEX,
+  MIDDLE_ENDIAN,
+  MIDDLE_ENDIAN_REGEX,
+  SEPARATOR_TEXT_DEFAULT,
+  USER_INPUT_REGEX,
+} from '../daterange.constants';
+import { NullableDaterange } from '../interfaces/daterange.interface';
+import { DayModel, NullableDayModel } from '../models/day.model';
+
+/**
+ * Daterange parsing service.
+ * Heavy inspired by `date-io.service.ts` from Clarity.
+ * https://github.com/vmware-clarity/ng-clarity/blob/a21fe726fe91ae3c929b8f016359afe354bb89c6/projects/angular/src/forms/datepicker/providers/date-io.service.ts
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class DaterangeParsingService {
+  public readonly localeDisplayFormat: InputDateDisplayFormat;
+  public readonly localeDelimiter: Array<string> = [];
+
+  /** Locale format. */
+  public get localeFormat(): string {
+    return (
+      this.localeDisplayFormat.format[0] +
+      this.localeDelimiter[0] +
+      this.localeDisplayFormat.format[1] +
+      this.localeDelimiter[1] +
+      this.localeDisplayFormat.format[2]
+    );
+  }
+
+  public constructor(@Inject(LOCALE_ID) private readonly locale: string) {
+    const localeFormat = getLocaleDateFormat(this.locale, FormatWidth.Short);
+    const format: string = localeFormat.toLocaleLowerCase();
+
+    console.log('DaterangeParsingService.getLocaleDisplayFormat', {
+      localeFormat,
+      format,
+    });
+
+    if (LITTLE_ENDIAN_REGEX.test(format)) {
+      this.localeDisplayFormat = LITTLE_ENDIAN;
+    } else if (MIDDLE_ENDIAN_REGEX.test(format)) {
+      this.localeDisplayFormat = MIDDLE_ENDIAN;
+    } else {
+      this.localeDisplayFormat = BIG_ENDIAN;
+    }
+
+    // Splitting it by date, month and year, leaves us with the delimiters.
+    // E.g. `"dd/MM/y".split(/d+|m+|y+/i)` results in `["", "/", "/", ""]`.
+    const delimiters: Array<string> = localeFormat.split(DELIMITER_REGEX);
+    this.localeDelimiter = [delimiters[1], delimiters[2]];
+  }
+
+  /**
+   * Convert daterange to locale friendly text.
+   * @param daterange - Daterange.
+   * @param separator - Separator between 'from' and 'to' dates.
+   * @returns Friendly daterange text.
+   */
+  public toLocaleString(daterange: NullableDaterange, separator: string = SEPARATOR_TEXT_DEFAULT): string {
+    console.log('DaterangeParsingService.toLocaleString', {
+      daterange,
+      separator,
+      locale: this.locale,
+    });
+
+    if (daterange == null) return '';
+
+    const { from, to } = daterange;
+    if (from == null || to == null) return '';
```

Comment:
Yeah, this is part of the discussion (as mentioned in the todo) of how to visually tackle this situation.

### Do we allow partial ranges (only 'from' or only 'to')?

* If yes, how do we visualize that? I like the proposed design [here](https://github.com/porscheinformatik/clarity-addons/issues/1887#issuecomment-1456703906), but that by itself would be a feature to build. Or could we re-use the time-control for this?
* If no, I'm fine with that. My companies use-case didn't require this. I can rebuild the logic to make it so that both values should always be required.

Reply:
@d-m-s What do you think of this one?

Reply:
I've been playing around with showing the date locale format where date is missing:

![image](https://user-images.githubusercontent.com/55841/227319764-b9f45d99-f0cc-4dd4-9f43-bf695eecb655.png)

Output from this example:
```json
  "daterange": {
    "to": {
      "year": 2023,
      "month": 2,
      "date": 23
    }
  }
```

Reply:
Alright we discussed how this should be handled from our point of view and we are fine going with your current solution. Our main problem with partially filled stuff was that we can't find a usable way for the manual input anymore.
But to be consistent, can you prevent a model update until the date is valid please?